trie/hexary.py

```diff
@@ -81,62 +85,95 @@ def _get(self, node, trie_key):
         else:
             raise Exception("Invariant: This shouldn't ever happen")
 
+    def _raise_missing_node(self, exception, key):
+        # Indicate more information about which key was requested, which node was missing, etc
+        raise MissingTrieNode(exception.args[0], self.root_hash, key) from exception
+
     def set(self, key, value):
         validate_is_bytes(key)
         validate_is_bytes(value)
 
         trie_key = bytes_to_nibbles(key)
-        root_node = self.get_node(self.root_hash)
 
-        new_node = self._set(root_node, trie_key, value)
+        try:
+            root_node = self.get_node(self.root_hash)
+
+            new_node = self._set(root_node, trie_key, value)
+        except KeyError as exc:
+            self._raise_missing_node(exc, key)
+
         self._set_root_node(new_node)
 
     def _set(self, node, trie_key, value):
-        self._prune_node(node)
-
         node_type = get_node_type(node)
 
+        # node is mutable, so capture the key for later pruning now
+        prune_key = self._get_prune_key(node)
```

Comment:
This seems like potential for forgetting, what about something like:

```python
with self._prune_node(node):
    ...  # do things that might raise errors
```

The pruning would only happen *if* the context exits without exception.

Reply:
Argh, how did I miss that. Thanks!